### PR TITLE
sql: add jsonb_exists_any operator

### LINKED ISSUE #54941: sql: add jsonb_exists_any operator
`jsonb_exists_any` is a builtin function that does the same thing as `?|` in Postgres. See the documentation here: https://www.postgresql.org/docs/current/functions-json.html

### PR DESCRIPTION
**Pull request by adzimzf**
Currently, CRDB doesn't have jsonb_exists_any operator.

Fixes #54941.

Release note (sql change): Implement the builtin `add jsonb_exists_any(jsonb, text[])` operator

### CODE REVIEW

**Review by yuzefovich — COMMENTED** [review]
Thanks for working on this!

Apart from having a builtin function, there is `?|` operator which does exactly the same thing, and I think it'd be great if we added the support for that operator in this PR as well. That change will be slightly more involved, but following an example of already existing operator (e.g. `JSONFetchVal`) should give you an idea of how to do it. Please let us know whether you'd be interested in working on that and whether more detailed instructions are necessary.

> Reviewed 3 of 3 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55172#-:-MIe9ckt4cd75q2k6FQ1:bd3k8i5)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adzimzf)

___
*[pkg/sql/logictest/testdata/logic_test/json_builtins, line 1153 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55172#-MIe7Ypw9AwCkE5K2R30:-MIe7Ypw9AwCkE5K2R31:bc1m8o6) ([raw file](https://github.com/cockroachdb/cockroach/blob/252d6da1ad9847c141ceb526c11ab0a9b70a0e2c/pkg/sql/logictest/testdata/logic_test/json_builtins#L1153)):*
> ```
> query BBBB
> select
>        jsonb_exists_any('{"id":12,"name":"Micheal","address": {"postcode":12,"state":"California"}}'::jsonb, array['id']),
> ```

super nit: I think you meant `Michael` and not `Micheal`.
___
*[pkg/sql/sem/builtins/builtins.go, line 2998 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55172#-MIe8l5547pTUJCAepAe:-MIe8l5547pTUJCAepAf:b-4lgpms) ([raw file](https://github.com/cockroachdb/cockroach/blob/252d6da1ad9847c141ceb526c11ab0a9b70a0e2c/pkg/sql/sem/builtins/builtins.go#L2998)):*
> ```Go
> 				j := tree.MustBeDJSON(args[0])
> 
> 				if j.Type() == json.ArrayJSONType {
> ```

Can you add a logic test for this case?
___
*[pkg/sql/sem/builtins/builtins.go, line 3019 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55172#-MIe8s6J7pg1mKd82le5:-MIe8s6J7pg1mKd82le6:b-gu8akp) ([raw file](https://github.com/cockroachdb/cockroach/blob/252d6da1ad9847c141ceb526c11ab0a9b70a0e2c/pkg/sql/sem/builtins/builtins.go#L3019)):*
> ```Go
> 				return tree.DBoolFalse, nil
> 			},
> 			Info:       "Determine weather any keys are exist in the JSON",
> ```

nit: some grammar errors, I'd say `Returns whether any of the strings in the text array exist as top-level keys or array elements` (based on Postgres docs).

Also, I wonder what "array elements" refers to, do you have an idea? Can you come up with a test query for this?

**Review by otan — CHANGES_REQUESTED** [review]
If it helps, I think the `?!` might be a separate issue/PR.
This current PR has a few improvements in functionality that we can make to make it compatible with Postgres.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55172#-:-MIeNDlX1iAVa4SYj4TL:b-4dswqs)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adzimzf and @yuzefovich)

___
*[pkg/sql/sem/builtins/builtins.go, line 2998 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55172#-MIe8l5547pTUJCAepAe:-MIeMYAG7LJV-IQiuIjk:b-rpd210) ([raw file](https://github.com/cockroachdb/cockroach/blob/252d6da1ad9847c141ceb526c11ab0a9b70a0e2c/pkg/sql/sem/builtins/builtins.go#L2998)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Can you add a logic test for this case?
</blockquote></details>

this case should return true if the jsonb is an array.

psql:
```

otan=# select jsonb_exists_any('["a"]', '{"a"}');
 jsonb_exists_any
------------------
 t
(1 row)
```

note the array can contains ints:

```
otan=# select jsonb_exists_any('["a", 10, 12]', '{"a"}'::text[]);
 jsonb_exists_any
------------------
 t
(1 row)
```
___
*[pkg/sql/sem/builtins/builtins.go, line 3004 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55172#-MIeMST-9i0tO_QcQ52S:-MIeMST-9i0tO_QcQ52T:bafojrs) ([raw file](https://github.com/cockroachdb/cockroach/blob/252d6da1ad9847c141ceb526c11ab0a9b70a0e2c/pkg/sql/sem/builtins/builtins.go#L3004)):*
> ```Go
> 				dArray := *tree.MustBeDArray(args[1])
> 				if dArray.Len() == 0 {
> 					return nil, pgerror.New(pgcode.IndeterminateDatatype,
> ```

this should work with an empty array for the any argument.

psql:
```
otan=# select jsonb_exists_any('["a"]', '{}');
 jsonb_exists_any
------------------
 f
(1 row)
```
___
*[pkg/sql/sem/builtins/builtins.go, line 3019 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55172#-MIe8s6J7pg1mKd82le5:-MIeMinA1T2_CmPt-02s:bik3cat) ([raw file](https://github.com/cockroachdb/cockroach/blob/252d6da1ad9847c141ceb526c11ab0a9b70a0e2c/pkg/sql/sem/builtins/builtins.go#L3019)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

nit: some grammar errors, I'd say `Returns whether any of the strings in the text array exist as top-level keys or array elements` (based on Postgres docs).

Also, I wonder what "array elements" refers to, do you have an idea? Can you come up with a test query for this?
</blockquote></details>

Please all terminate with a full stop.

**Review by adzimzf — COMMENTED** [review]
I'd be happy to work the operator `?!`,
should I add it to different PR or the same?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55172#-:-MIlV-WU4r5erfoqOcXj:bmmdfzs)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan and @yuzefovich)

___
*[pkg/sql/logictest/testdata/logic_test/json_builtins, line 1153 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55172#-MIe7Ypw9AwCkE5K2R30:-MIlV_944d_RxKOjhll6:bytjdte) ([raw file](https://github.com/cockroachdb/cockroach/blob/252d6da1ad9847c141ceb526c11ab0a9b70a0e2c/pkg/sql/logictest/testdata/logic_test/json_builtins#L1153)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

super nit: I think you meant `Michael` and not `Micheal`.
</blockquote></details>

sorry for the typo
___
*[pkg/sql/sem/builtins/builtins.go, line 2998 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55172#-MIe8l5547pTUJCAepAe:-MIlSAvz5HGmS5oKkduY:b45z71h) ([raw file](https://github.com/cockroachdb/cockroach/blob/252d6da1ad9847c141ceb526c11ab0a9b70a0e2c/pkg/sql/sem/builtins/builtins.go#L2998)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

this case should return true if the jsonb is an array.

psql:
```

otan=# select jsonb_exists_any('["a"]', '{"a"}');
 jsonb_exists_any
------------------
 t
(1 row)
```

note the array can contains ints:

```
otan=# select jsonb_exists_any('["a", 10, 12]', '{"a"}'::text[]);
 jsonb_exists_any
------------------
 t
(1 row)
```
</blockquote></details>

noted, 
on it to add the logical test and update the code.
___
*[pkg/sql/sem/builtins/builtins.go, line 3004 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55172#-MIeMST-9i0tO_QcQ52S:-MIlVnu1CDDMUOLzs4Dv:b-4zhqd1) ([raw file](https://github.com/cockroachdb/cockroach/blob/252d6da1ad9847c141ceb526c11ab0a9b70a0e2c/pkg/sql/sem/builtins/builtins.go#L3004)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

this should work with an empty array for the any argument.

psql:
```
otan=# select jsonb_exists_any('["a"]', '{}');
 jsonb_exists_any
------------------
 f
(1 row)
```
</blockquote></details>

Sorry I didn't notice there will be a case like this.
Any hints on how to spot the argument are `'{}'` or `array[]` ?, because in Postgres when the argument is `array[]` it'll throw an error ` Cannot determine type of empty array`

here's the query
```sql
    select jsonb_exists_any('{"id":12,"name":"Micheal","address": {"postcode":12,"state":"California"}}'::jsonb, array[]);
```
___
*[pkg/sql/sem/builtins/builtins.go, line 3019 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55172#-MIe8s6J7pg1mKd82le5:-MIlWw3R4epRzdOZNw4o:b-7m2y43) ([raw file](https://github.com/cockroachdb/cockroach/blob/252d6da1ad9847c141ceb526c11ab0a9b70a0e2c/pkg/sql/sem/builtins/builtins.go#L3019)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

Please all terminate with a full stop.
</blockquote></details>

noted

**Review by adzimzf — COMMENTED** [review]
Hi @yuzefovich,
I try to run this query in CRDB,
```sql
select '{"id":12,"name":"Michael","address": {"postcode":12,"state":"California"}}'::jsonb ?| array['id'];
```
seems CRDB has had it


> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55172#-:-MIlby3f1GcLZnrw07iO:b-3bpqu4)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan and @yuzefovich)

**Comment by otan** [comment]
If it already exists we should re-use the same code.
https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/sem/tree/eval.go#L2408

**Comment by adzimzf** [comment]
> If it already exists we should re-use the same code.
> https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/sem/tree/eval.go#L2408

noted, I'll reuse the existing code.
@otan regarding the empty array, I check the operator `?|` has a different result compared to Postgres.
image

picture 1 from Postgres and picture 2 from CRDB,
is it okay?

**Comment by otan** [comment]
Difference in this case is ok

**Comment by adzimzf** [comment]
> Difference in this case is ok

well and noted

File: pkg/sql/sem/tree/eval.go [hunk 1]
```diff
@@ -383,6 +383,24 @@ func ArrayContains(ctx *EvalContext, haystack *DArray, needles *DArray) (*DBool,
 	return DBoolTrue, nil
 }
 
+// JSONExistAny return true if any value in dArray is exist in the json
+func JSONExistAny(_ *EvalContext, json DJSON, dArray *DArray) (*DBool, error) {
```
[the remark this answers is not in the document]
Comment (adzimzf) [reply pkg/sql/sem/tree/eval.go:387]: sure

[the remark this answers is not in the document]
Comment (adzimzf) [reply pkg/sql/sem/tree/eval.go]: done

**Review by otan — APPROVED** [review]

**Comment by otan** [comment]
thanks!

bors r+

**Pull request merged**